cram_urdf_projection/cram_urdf_projection/src/low-level.lisp

```diff
@@ -0,0 +1,583 @@
+     cram-tf:*fixed-frame*
+     cram-tf:*robot-base-frame*
+     (cut:current-timestamp)
+     pose-in-map)))
+
+;;;;;;;;;;;;;;;;; NAVIGATION ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defun drive (target)
+  (declare (type cl-transforms-stamped:pose-stamped target))
+  (let* ((world btr:*current-bullet-world*)
+         (world-state (btr::get-state world)))
+    (unwind-protect
+         (assert
+          (prolog:prolog
+           `(and (cram-robot-interfaces:robot ?robot)
+                 (btr:bullet-world ?w)
+                 (btr:assert ?w (btr:object-pose ?robot ,target)))))
+      (when (btr:robot-colliding-objects-without-attached '(:floor))
+        (unless (< (abs *debug-short-sleep-duration*) 0.0001)
+          (cpl:sleep *debug-short-sleep-duration*))
+        (btr::restore-world-state world-state world)
+        (cpl:fail 'common-fail:navigation-pose-unreachable :pose-stamped target)))))
+
+;;;;;;;;;;;;;;;;; TORSO ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(defun move-torso (joint-angle)
+  (declare (type number joint-angle))
+  (let* ((bindings
+           (car
+            (prolog:prolog
+             `(and (cram-robot-interfaces:robot ?robot)
+                   (btr:bullet-world ?w)
+                   (cram-robot-interfaces:robot-torso-link-joint ?robot ?_ ?joint)
+                   (cram-robot-interfaces:joint-lower-limit ?robot ?joint ?lower)
+                   (cram-robot-interfaces:joint-upper-limit ?robot ?joint ?upper)))))
+         (lower-limit
+           (cut:var-value '?lower bindings))
+         (upper-limit
+           (cut:var-value '?upper bindings))
+         (cropped-joint-angle
+           (if (< joint-angle lower-limit)
+               lower-limit
+               (if (> joint-angle upper-limit)
+                   upper-limit
+                   joint-angle))))
+    (prolog:prolog
+     `(btr:assert (btr:joint-state ?w ?robot ((?joint ,cropped-joint-angle))))
+     bindings)
+    (unless (< (abs (- joint-angle cropped-joint-angle)) 0.0001)
+      (cpl:fail 'common-fail:torso-goal-not-reached
+                :description (format nil "Torso goal ~a was out of joint limits" joint-angle)
+                :torso joint-angle))))
+
+;;;;;;;;;;;;;;;;; PTU ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+
+(defun look-at-joint-angles (joint-angles)
+  (declare (type list joint-angles))
+  (assert
+   (prolog:prolog
+    `(and (cram-robot-interfaces:robot ?robot)
+          (btr:bullet-world ?w)
+          (cram-robot-interfaces:robot-pan-tilt-joints ?robot . ?joint-names)
+          (prolog:lisp-fun mapcar list ?joint-names ,joint-angles ?joint-states)
+          (btr:assert ?w (btr:joint-state ?robot ?joint-states))))))
+
+(defun look-at-joint-states (joint-states)
+  (declare (type list joint-states))
+  (assert
+   (prolog:prolog
+    `(and (cram-robot-interfaces:robot ?robot)
+          (btr:bullet-world ?w)
+          (btr:assert ?w (btr:joint-state ?robot ,joint-states))))))
+
+(defun look-at (pose configuration)
```

Comment:
It would be more consistent if ``look-at`` came after all the 3 variations of look, i.e. after ``look-at-pose-stamped``, please change the places of ``look-at`` and ``look-at-pose-stamped``.